c2rust-analyze/src/rewrite/expr/distribute.rs

```diff
@@ -0,0 +1,126 @@
+use crate::rewrite::expr::mir_op::{self, MirRewrite, SubLoc};
+use crate::rewrite::expr::unlower::{MirOrigin, MirOriginDesc};
+use log::*;
+use rustc_hir::HirId;
+use rustc_middle::mir::Location;
+use rustc_middle::ty::TyCtxt;
+use std::collections::{BTreeMap, HashMap};
+
+struct RewriteInfo {
+    rw: mir_op::RewriteKind,
+    loc: Location,
+    sub_loc: Vec<SubLoc>,
+    desc: MirOriginDesc,
+}
+
+/// Distributes MIR rewrites to HIR nodes.  This takes a list of MIR rewrites (from `mir_op`) and a
+/// map from MIR location to `HirId` (from `unlower`) and produces a map from `HirId` to a list of
+/// MIR rewrites.
+///
+/// Using the example from `unlower`:
+///
+/// ```text
+/// bb0[5]: Terminator { source_info: ..., kind: _4 = f(move _5) -> [return: bb1, unwind: bb2] }
+///   []: StoreIntoLocal, `f(y)`
+///   [Rvalue]: Expr, `f(y)`
+///   [Rvalue, CallArg(0)]: Expr, `y`
+/// ```
+///
+/// A MIR rewrite on `bb0[5]` `[Rvalue, CallArg(0)]` would be attached to the MIR
+/// `Expr` `y`, and a rewrite on `bb0[5]` `[Rvalue]` would be attached to `f(y)`.
+/// A MIR rewrite on `bb0[5]` `[]` (i.e. on the call terminator itself) would
+/// result in an error, since there is no good place in the HIR to attach such a
+/// rewrite.
+pub fn distribute<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    unlower_map: BTreeMap<(Location, Vec<SubLoc>), MirOrigin>,
+    mir_rewrites: HashMap<Location, Vec<MirRewrite>>,
+) -> HashMap<HirId, Vec<mir_op::RewriteKind>> {
+    let mut info_map = HashMap::<HirId, Vec<RewriteInfo>>::new();
+
+    for (loc, mir_rws) in mir_rewrites {
+        for mir_rw in mir_rws {
+            let mut key = (loc, mir_rw.sub_loc);
```

Comment:
I think adding a

```rust
struct PreciseLoc {
    loc: Location,
    sub_loc: Vec<SubLoc>,
}
```

would make this a lot more readable (`key.1` isn't very) (and I think we may also be able to `impl Borrow<Q> for PreciseLoc` to avoid `sub_loc` clones, though I'm not sure that's worth the complexity).

Reply:
Added a `struct PreciseLoc`.  `impl Borrow` I don't think is possible because it needs to return `&Q` where `Q` is something like `(Location, &[SubLoc])`, but `PreciseLoc` can't own such a thing.